File: src/services/guestService.ts
/**
 * Additional guests sharing a booking's room.
 *
 * Recording companions is optional: a booking with none behaves exactly as it
 * did before this existed. What the service guarantees is that a booking never
 * holds more companions than the room was booked for, and that a companion's ID
 * photos are cleaned up whenever the companion or the booking goes away.
 */

import { nowIso } from '../utils/date';
import { AppError } from '../utils/errors';
import {
  normaliseGuestInput,
  validateGuestForm,
  hasErrors,
} from '../utils/validation';
import type { BookingGuest, BookingGuestInput, GuestCapacity, PreparedImage } from '../types';
import type { BookingGuestRepository } from '../database/repositories/bookingGuestRepository';
import type { CustomerRepository } from '../database/repositories/customerRepository';
import type { ImageService } from './imageService';

export interface GuestImages {
  front?: PreparedImage;
  back?: PreparedImage;
}

export class GuestService {
  constructor(
    private readonly guests: BookingGuestRepository,
    private readonly customers: CustomerRepository,
    private readonly images: ImageService,
  ) {}

  async list(customerId: number): Promise<BookingGuest[]> {
    return this.guests.listByCustomer(customerId);
  }

  async get(guestId: number): Promise<BookingGuest> {
    const guest = await this.guests.findById(guestId);
    if (!guest) throw new AppError('NOT_FOUND', 'That guest record no longer exists.');
    return guest;
  }

  /**
   * How many companions this booking has room for. The primary guest counts as
   * one of `number_of_persons`, so a booking for 3 allows 2 companions.
   */
  async capacity(customerId: number): Promise<GuestCapacity> {
    const customer = await this.customers.findById(customerId);
    if (!customer) throw new AppError('NOT_FOUND', 'That customer record no longer exists.');
    const recorded = await this.guests.countByCustomer(customerId);
    const allowed = Math.max(0, customer.number_of_persons - 1);
    return {
      persons: customer.number_of_persons,
      recorded,
      allowed,
      canAddMore: recorded < allowed,
    };
  }

  /** Adds a companion, optionally with their own ID photos. */
  async add(
    customerId: number,
    values: BookingGuestInput,
    images: GuestImages = {},
  ): Promise<BookingGuest> {
    const errors = validateGuestForm(values);
    if (hasErrors(errors)) {
      throw new AppError('VALIDATION', Object.values(errors).find(Boolean) ?? 'Please check the form.');
    }

    const customer = await this.customers.findById(customerId);
    if (!customer) throw new AppError('NOT_FOUND', 'That customer record no longer exists.');

    const capacity = await this.capacity(customerId);
    if (!capacity.canAddMore) {
      throw new AppError(
        'VALIDATION',
        capacity.allowed === 0
          ? 'This booking is for one person. Increase "Number of Persons" to record other guests.'
          : `This booking is for ${capacity.persons} persons, so only ${capacity.allowed} other ` +
            'guests can be recorded. Increase "Number of Persons" to add more.',
      );
    }

    const input = normaliseGuestInput(values);
    const timestamp = nowIso();
    // The row goes in first so its id can name the photo folder.
    const id = await this.guests.insert({
      customer_id: customerId,
      name: input.name,
      phone: input.phone,
      created_at: timestamp,
      updated_at: timestamp,
    });

    try {
      await this.storePhotos(customer.customer_code, id, images, timestamp);
    } catch (e) {
      // Roll the row back so a failed photo write leaves no half-record.
      await this.guests.delete(id);
      await this.images.removeGuestImages(customer.customer_code, id);
      throw e;
    }

    return this.get(id);
  }

  /** Updates a companion's details, and replaces any photo that was retaken. */
  async update(
    guestId: number,
    values: BookingGuestInput,
    images: GuestImages = {},
  ): Promise<BookingGuest> {
    const errors = validateGuestForm(values);
    if (hasErrors(errors)) {
      throw new AppError('VALIDATION', Object.values(errors).find(Boolean) ?? 'Please check the form.');
    }

    const guest = await this.get(guestId);
    const customer = await this.customers.findById(guest.customer_id);
    if (!customer) throw new AppError('NOT_FOUND', 'That customer record no longer exists.');

    const input = normaliseGuestInput(values);
    const timestamp = nowIso();
    await this.guests.update(guestId, {
      name: input.name,
      phone: input.phone,
      updated_at: timestamp,
    });
    // Replacing a photo overwrites the same filename, so no orphan is created.
    await this.storePhotos(customer.customer_code, guestId, images, timestamp);

    return this.get(guestId);
  }

  /** Removes a companion and their photos. */
  async remove(guestId: number): Promise<void> {
    const guest = await this.get(guestId);
    const customer = await this.customers.findById(guest.customer_id);

    const changes = await this.guests.delete(guestId);
    if (changes === 0) {
      throw new AppError('DB_QUERY', 'The guest record could not be deleted.');
    }
    // The row is gone first, so a filesystem hiccup cannot leave a record
    // pointing at deleted photos.
    if (customer) {
      await this.images.removeGuestImages(customer.customer_code, guestId);
    }
  }

  private async storePhotos(
    customerCode: string,
    guestId: number,
    images: GuestImages,
    timestamp: string,
  ): Promise<void> {
    const patch: Parameters<BookingGuestRepository['update']>[1] = { updated_at: timestamp };
    let touched = false;

    if (images.front) {
      const stored = await this.images.storeGuestIdImage(customerCode, guestId, 'front', images.front);
      patch.id_front_path = stored.path;
      patch.id_front_thumb_path = stored.thumbPath;
      touched = true;
    }
    if (images.back) {
      const stored = await this.images.storeGuestIdImage(customerCode, guestId, 'back', images.back);
      patch.id_back_path = stored.path;
      patch.id_back_thumb_path = stored.thumbPath;
      touched = true;
    }
    if (touched) await this.guests.update(guestId, patch);
  }
}

File: lib/services/guest_service.dart
import '../database/repositories/booking_guest_repository.dart';
import '../database/repositories/customer_repository.dart';
import '../models/models.dart';
import '../utils/app_date.dart';
import '../utils/app_error.dart';
import '../utils/validation.dart';
import 'image_service.dart';

class GuestImages {
  final PreparedImage? front;
  final PreparedImage? back;
  const GuestImages({this.front, this.back});
}

/// Companion guests sharing a room with the primary booking.
class GuestService {
  final BookingGuestRepository _guests;
  final CustomerRepository _customers;
  final ImageService _images;
  GuestService(this._guests, this._customers, this._images);

  Future<List<BookingGuest>> list(int customerId) => _guests.listByCustomer(customerId);

  Future<BookingGuest> get(int guestId) async {
    final guest = await _guests.findById(guestId);
    if (guest == null) {
      throw const AppError(AppErrorCode.notFound, 'That guest record was not found.');
    }
    return guest;
  }

  Future<GuestCapacity> capacity(int customerId) async {
    final customer = await _customers.findById(customerId);
    if (customer == null) {
      throw const AppError(AppErrorCode.notFound, 'That customer record was not found.');
    }
    final allowed = customer.numberOfPersons - 1 < 0 ? 0 : customer.numberOfPersons - 1;
    final recorded = await _guests.countByCustomer(customerId);
    return GuestCapacity(
      persons: customer.numberOfPersons,
      recorded: recorded,
      allowed: allowed,
      canAddMore: recorded < allowed,
    );
  }

  Future<BookingGuest> add(int customerId, NormalisedGuestForm values, [GuestImages images = const GuestImages()]) async {
    final customer = await _customers.findById(customerId);
    if (customer == null) {
      throw const AppError(AppErrorCode.notFound, 'That customer record was not found.');
    }
    final cap = await capacity(customerId);
    if (!cap.canAddMore) {
      final message = cap.allowed == 0
          ? 'This booking is for one person, so no other guests can be recorded. Increase '
              '"Number of Persons" on the booking first.'
          : 'This booking already has all ${cap.allowed} other guest(s) recorded. Increase '
              '"Number of Persons" on the booking to add more.';
      throw AppError(AppErrorCode.validation, message);
    }
    final now = nowIso();
    final guestId = await _guests.insert(NewBookingGuestRow(
      customerId: customerId,
      name: values.name,
      phone: values.phone,
      createdAt: now,
      updatedAt: now,
    ));
    try {
      await _storePhotos(customer.customerCode, guestId, images, now);
    } catch (e) {
      await _guests.delete(guestId);
      await _images.removeGuestImages(customer.customerCode, guestId);
      rethrow;
    }
    return get(guestId);
  }

  Future<BookingGuest> update(int guestId, NormalisedGuestForm values, [GuestImages images = const GuestImages()]) async {
    final guest = await get(guestId);
    final customer = await _customers.findById(guest.customerId);
    if (customer == null) {
      throw const AppError(AppErrorCode.notFound, 'That customer record was not found.');
    }
    final now = nowIso();
    await _guests.update(guestId, {
      'name': values.name,
      'phone': values.phone,
      'updated_at': now,
    });
    await _storePhotos(customer.customerCode, guestId, images, now);
    return get(guestId);
  }

  Future<void> remove(int guestId) async {
    final guest = await get(guestId);
    final customer = await _customers.findById(guest.customerId);
    final changes = await _guests.delete(guestId);
    if (changes == 0) {
      throw const AppError(AppErrorCode.notFound, 'That guest record was not found.');
    }
    if (customer != null) {
      await _images.removeGuestImages(customer.customerCode, guestId);
    }
  }

  Future<void> _storePhotos(String customerCode, int guestId, GuestImages images, String now) async {
    final patch = <String, Object?>{};
    if (images.front != null) {
      final stored = await _images.storeGuestIdImage(customerCode, guestId, PhotoSide.front, images.front!);
      patch['id_front_path'] = stored.path;
      patch['id_front_thumb_path'] = stored.thumbPath;
    }
    if (images.back != null) {
      final stored = await _images.storeGuestIdImage(customerCode, guestId, PhotoSide.back, images.back!);
      patch['id_back_path'] = stored.path;
      patch['id_back_thumb_path'] = stored.thumbPath;
    }
    if (patch.isNotEmpty) {
      patch['updated_at'] = now;
      await _guests.update(guestId, patch);
    }
  }
}
